### PR TITLE
Pipes - Kimberley Zell - Api-Muncher

### PR DESCRIPTION
**Pull request by kimpossible1**
# API Muncher
Congratulations! You're submitting your assignment!

## Comprehension Questions
| Question | Answer    |
| :------------- | :------------- |
| How did you go about exploring the Edamam API, how did you try querying the API? | I used Postman and then the console to check the parsed response for my expectations |
| Describe your API Wrapper.  How did you decide on the methods you created? | I followed along with the requirements. I created a self.search method that returns a list of recipes according to entry from user. I also made a self.find method that gets a specific recipe from the api, based on the uri of the recipe. I have a helper method in private methods that creates the individual recipe instances with only the data I have listed. |
| Describe an edge case or failure case test you wrote for your API Wrapper. | I checked that it returns nil if a recipe uri is invalid or not found. |
| Explain how VCR aids in testing an API. | It records data from a test query so that future tests on same query can use that saved information rather than query the API repeatedly  |
| What is the Heroku URL of your deployed application? |https://recipe-monster.herokuapp.com/  |

### CODE REVIEW

File: app/views/layouts/application.html.erb [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+        <section class="hero">
+          <h1>
+            <%= link_to "Recipe Monster", root_path %>
+          </h1>
+
+        </section>
+    </header>
+
+    <section class="main">
+      <% if flash[:message] %>
```
Comment (droberts-sea) [inline app/views/layouts/application.html.erb:29]: Why `<section class="main">` instead of `<main>`?

File: lib/recipe_api_wrapper.rb [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+  APP_ID = ENV["APP_ID"]
+  APP_KEY = ENV["APP_KEY"]
+
+  def self.search(food, from)
+    url = BASE_URL + "?q=#{food}" + "&from=#{from}" "&app_id=#{APP_ID}" + "&app_key=#{APP_KEY}"
+
+    response = self.parse_response(HTTParty.get(url))
+    recipe_list = []
+
+    if response.is_a?(Hash) && response["hits"]
```
Comment (droberts-sea) [inline lib/recipe_api_wrapper.rb:19]: I like these checks - they'll definitely keep your program form trying to read data that doesn't exist.

One thing I'm noticing: this method doesn't have any way to distinguish between an error response (for example because of bad API keys) and a success response with no hits.

Since one of these cases is a programer error and the other indicates the user did something wrong, I would argue that they ought to be handled differently. In order to do so, this method would have to return some more information.

File: test/controllers/recipes_controller_test.rb [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+
+    it "returns a collection of recipes when from parameter is given" do
+      VCR.use_cassette("recipes") do
+        get root_path, params: { search: "cheese", from: 10 }
+        must_respond_with :success
+      end
+    end
+
+    it "returns a collection of recipes when from parameter is not given" do
+      VCR.use_cassette("recipes") do
```
Comment (droberts-sea) [inline test/controllers/recipes_controller_test.rb:20]: I like that you exercise both these cases around the `from` parameter. Other cases I'd be interested in seeing here:

1. What if the search returns no results (for example, a search for "dfaljdfalflaksjflaksj")?
1. What if the paging parameters are invalid (negative number, etc)?
1. What if the paging parameters take you past the end of the available search results?

The first of those is covered by your lib tests below, but it's still important to verify the controller does the right thing in this situation.

File: test/controllers/recipes_controller_test.rb [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+    it "returns success when a valid uri is given" do
+      VCR.use_cassette("specific_recipe") do
+        get recipes_path, params: { id: "http://www.edamam.com/ontologies/edamam.owl#recipe_421df807e21c65c842ec62870604aed3" }
+        must_respond_with :success
+      end
+    end
+
+    it "returns not found when an invalid uri is given" do
+      VCR.use_cassette("bad_uri_recipe") do
+        get recipes_path, params: { id: "http://www.edamam.com/ontologies/edamam.owl#recipe_421df807e21c65c842ec62870604" } #bad uri
```
Comment (droberts-sea) [inline test/controllers/recipes_controller_test.rb:37]: Good failure test

File: test/lib/recipe_api_wrapper_test.rb [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+
+  it "can find a specific recipe" do
+    uri = ""
+    VCR.use_cassette("recipes") do
+      result = RecipeApiWrapper.search("cheese", 10)
+      uri = result[0].id
+    end
+
+    VCR.use_cassette("recipe") do
+      recipe = RecipeApiWrapper.find(uri)
```
Comment (droberts-sea) [inline test/lib/recipe_api_wrapper_test.rb:31]: I like the pattern of this test - using real search results to seed the call to `find`.

File: test/lib/recipe_api_wrapper_test.rb [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+    end
+  end
+
+  it "does returns nil when the API is not working" do
+    VCR.use_cassette("no_recipes") do
+      result = RecipeApiWrapper.search("kljkjlk", 10)
+      result.must_be_kind_of Array
+      result.must_be_empty
+    end
+  end
```
Comment (droberts-sea) [inline test/lib/recipe_api_wrapper_test.rb:21]: I would like to see some testing around the second parameter to `search` here.

**Comment by droberts-sea** [comment]
# API Muncher
## What We're Looking For

| Feature | Feedback    |
| :------------- | :------------- |
| **Core Requirements** |   |
| Git hygiene | yes |
| Comprehension questions	| yes |
| **General** |  |
| Rails fundamentals (RESTful routing, use of named paths) | yes |
| Semantic HTML | yes |
| Errors are reported to the user | yes |
| API Wrapper to handle the API requests | yes |
| Controller testing | mostly - see inline comments |
| Lib testing | mostly - see inline comments |
| Search Functionality | yes |
| List Functionality | yes |
| Show individual item functionality (link to original recipe opens in new tab) | yes |
| **Styling** |  |
| Foundation Styling for responsive layout | yes |
| List View shows 10 items at a time/pagination | yes |
| The app is styled to create an attractive user interface | yes - In some places the text was hard to read on top of the background. One strategy to get around this is to make the element containing the text slightly opaque to give better contrast. |
| **API Features** | |
| The App attributes Edaman | yes |
| The VCR casettes do not contain the API key | yes |
| **External Resources** |  |
| Link to deployed app on Heroku | yes |
|  **Overall** | Good job overall! The core functionality of the app (querying an API and turning the results into a webpage) works great. Error handling isn't as fine-grained as I might like and (as with most software projects) there's some room to expand your test coverage, but in general I'm quite happy with this submission. Keep up the hard work! |
